_posts/2018-11-29-async-update.md

```diff
@@ -0,0 +1,201 @@
+community has made quite a lot of progress toward that goal. This post
+summarizes the state of play, and announces the publication several crates
+intended to facilitate use of async/await on the nightly ecosystem.
+
+# Why async/await
+
+Before delving into the current status, it's worth taking a moment to recap
+the core motivations for async/await, and its special importance for Rust.
+
+Async/await notation is a way of making asynchronous programming more
+closely resemble synchronous programming. To see how this works, consider
+`Read::read` in `std::io`:
+
+```rust
+fn read(&mut self, buf: &mut [u8]) -> Result<usize, std::io::Error>
+```
+
+This synchronous method *blocks the current thread* until data has been read
+into `buf`, then says how many bytes were read. We can build on this method
+to implement `read_exact`, a method that *continues* reading until the buffer
+is filled:
+
+```rust
+fn read_exact<T: Read>(input: &mut T, buf: &mut [u8]) -> Result<(), std::io::Error> {
+    let mut cursor = 0;
+    while cursor < buf.len() {
+        cursor += input.read(&mut buf[cursor..])?;
+    }
+}
+```
+
+In the asynchronous world, we want to perform similar operations, but rather
+than blocking the current thread we want to leave it free to do other work
+while the I/O operations complete asychronously. But actually programming
+*directly* in that way is incredibly difficult. What we want is to program
+*as if* I/O operations will block the current thread, but have supporting
+infrastructure transform this code into more efficient asynchronous execution.
+
+In short, our goal is to be able to write the following:
+
+```rust
+async fn read_exact<T: AsyncRead>(input: &mut T, buf: &mut [u8]) -> Result<(), std::io::Error> {
+    let mut cursor = 0;
+    while cursor < buf.len() {
+        cursor += await!(input.read(&mut buf[cursor..]))?;
+    }
+}
+```
+
+Comparing the two snippets, there are three changes here:
+
+- We write `async` before `fn`, to signal that the function should be
+asynchronously executed on its parent thread. Async functions return
+their result within a `Future`, representing a value that must be asynchronously
+computed.
+
+- We use the `AsyncRead` trait (from `futures::io`), rather than the
+`Read` trait. This makes an asynchronous version of the `read` method
+available (curretly provided via the `AsyncReadExt` extension trait).
+
+- We enclose the call to `read` with `await!`, signaling that we want to
+*simulate* blocking on the operation to complete.
+
+And that's all.
+
+This kind of systematic transformation is one that has proved itself
+repeatedly in other languages. But there's an extra element in the Rust
+version: borrowing. For `read_exact`, we are able to hold the borrows
+of `input` and `buf` while we use `await!` (which may actually clear
+the stack and run completely unrelated code). The validity of the
+borrowing is still checked, and it works largely similarly to borrowing
```

Comment:
Is it just largely similar, or is it identical? I can't think of any discrepancies off the top of my head.